pilot/pkg/model/push_context.go

```diff
@@ -78,9 +80,86 @@ type PushContext struct {
 	// Env has a pointer to the shared environment used to create the snapshot.
 	Env *Environment `json:"-,omitempty"`
 
+	// ServicePorts is used to keep track of service name and port mapping.
+	// This is needed because ADS names use port numbers, while endpoints use
+	// port names. The key is the service name. If a service or port are not found,
+	// the endpoint needs to be re-evaluated later (eventual consistency)
+	ServiceName2Ports map[string]map[string]uint32 `json:"-"`
+	ServicePort2Name  map[string]map[uint32]string `json:"-"`
+
 	initDone bool
 }
 
+// EDSUpdater is used for direct updates of the EDS model
+// and push.
+type EDSUpdater interface {
+	// EDSUpdate is called when the list of endpoints or labels in a ServiceEntry is
+	// changed. For each cluster and hostname, the full list of active endpoints (including empty list)
+	// must be sent.
+	EDSUpdate(shard, hostname string, entry []*IstioEndpoint) error
+
+	// SvcUpdate is called when a service port mapping definition is updated.
+	// This interface is WIP - labels, annotations and other changes to service may be
+	// updated to force a EDS and CDS recomputation and incremental push, as it doesn't affect
+	// LDS/RDS.
+	SvcUpdate(shard, hostname string, ports map[string]uint32, rports map[uint32]string)
+}
+
+// IstioEndpoint has the information about a single address+port for a specific
+// service and shard. Unlike ServiceInstance, it has less info and require less
+// allocation.
+type IstioEndpoint struct {
+
+	// Labels points to the workload or deployment labels.
+	Labels map[string]string
+
+	// Address is the IP address of the endpoint. Other types may be
+	// supported in future.
+	Address string
+
+	// EndpointPort is the port where the workload is listening, can be different
+	// from the service port.
+	EndpointPort uint32
+
+	ServicePortName string
+
```

Comment:
Added the comments.
"// ServicePortName tracks the name of the port, to avoid 'eventual consistency' issues.
	// Sometimes the Endpoint is visible before Service - so looking up the port number would
	// fail. Instead the mapping to number is made when the clusters are computed. The lazy
	// computation will also help with 'on-demand' and 'split horizon' - where it will be skipped
	// for not used clusters or endpoints behind a gate."